fix(compiler) restore old copy behaviour, with new ignore option

George-Payne opened the pull request:
## What is the current behavior?

#5899 introduced a lot of regressions:

A breaking change in copy behavior: #5983
file to file copying broken: #5956
EBUSY: resource busy: #5966
Directories from above the root dir are copied outside of the output dir. 

## What is the new behavior?

This PR restores the old copy behavior, but uses the passed filter list, rather than the hardcoded one.

This isn't an optimum fix, as it changes the copy back to `dir` to `dir` (or `file` to `file`), so ignore patterns aren't matched against child files. Only glob matches are ignored (as was the previous behavior). IMO it's better than the large number of regressions, especially as this can now be worked around (e.g. in the case of #5781) by the user moving to a glob copy, with the new ignore option.

## Documentation

none

## Does this introduce a breaking change?

- [ ] Yes
- [x] No


## Testing

Tested on own repo.

## Other information

fixes: #5983
fixes: #5956
fixes: #5966

dutscher commented:
we encounter copy problems with `"@stencil/angular-output-target": "^0.9.0", "@stencil/react-output-target": "^0.5.3", ` after update to `4.21`.
i guess this restore will solve this issue.
cheers

Armand-Lluka commented:
Hey, is there any update or traction on this? This is currently acting as a blocker from us updating.

christian-bromann reviewed (approved):
LGTM 👍 

Sorry for this regression. We were not completely aware of all use cases. This will help us fix the current issues and eventually attempt to get this right a second time.

christian-bromann commented:
@George-Payne it seems like the copy task test I added is failing, mind checking?

George-Payne commented:
> @George-Payne it seems like the copy task test I added is failing, mind checking?

Didn't realise there were multiple test commands, so missed that one. I moved it to use a glob and amended the commit.

Armand-Lluka commented:
Hey, is there anything blocking this from being added?

christian-bromann reviewed (approved):
LGTM 👍

christian-bromann commented:
@George-Payne thanks for the contribution, we will release this next week on Monday.

cavez86 commented:
Hi @christian-bromann sorry to bother, there is any news on the release of this?

dutscher commented:
@christian-bromann in the v4.22.3 changelog isn't row of this merge, also in the commits section of this tag.
are you sure that this merge is in 4.22.3?
cheers and thanks in advance
